### ISSUE TITLE
Synchronization does not support proxy

### ISSUE THREAD

**Issue by gibboni** [laurent22/joplin#164]
### Operating system

- [X ] Windows 7
- [X ] Windows 10

### Application

- [X ] Desktop

### Error

The desktop-app does not support proxy or it fails to use system-wide proxy settings. When the synchronise-button is pressed, nothing seem to happen. There is no visible error message. The error, ECONNREFUSED, is however logged properly into the log file.

**Comment by laurent22** [comment]
I'm wondering, is it the application that fails to use the proxy, or OneDrive that is refusing connections via this proxy? Does the native OneDrive client work in this setup?

**Comment by gibboni** [comment]
Native OneDrive works just fine. Based on the log, the app tries to connect directly to graph.microsoft.com. Are there any environment or command line settings I could try?

**Comment by laurent22** [comment]
It seems Node indeed does not use the system proxy settings by default - https://stackoverflow.com/a/27825729/561309 So this will need to be implemented

**Comment by jogi4github** [comment]
Is there a way to tackle this on windows systems. On Linux I guess an export of http[s]_proxy will work - but what about Windows? I was not able to find a hack for the proxy issue here...

**Comment by artelse** [comment]
I get the same `reason: connect ECONNREFUSED` error when trying to connect to a local webdav server. Is there a workaround/solution?

**Comment by aksdb** [comment]
On Linux I also get ECONNREFUSED, even though I have http[s]_proxy set.

**Comment by crequill** [comment]
Same problem here on linux: with http[s]_proxy set, I get network timeout error with desktop application and nextcloud synchronization.

**Comment by cossio** [comment]
Same problem here. Trying to setup Dropbox sync, but it gets stuck in the "Enter the code provided by Dropbox" dialog.

I am on Linux, with $http_proxy set.

**Comment by PyroP** [comment]
Same problem here in Windows 10.

**Comment by think-nice-things** [comment]
Hi -- any update on this?
I'm on windows 7 and the system proxy isn't recognized :-(

**Comment by garzy** [comment]
same problem too. Windows version must have a dialog for configure proxy setttiongs.

Thanks!

**Comment by jogi4github** [comment]
Hello Jürgen,

I got it working by installing the NextCloud client, syncing the folder with the md files and configuring  a file system snyc in Joplin.
Not sure if this is the best way, but at least it works for me.

- Jogi

**Comment by ZZMajor** [comment]
Hi,

yeah, that would be an idea but just a bit of overkill. Instead of installing an further app to sync files it would be great if Joplin could handle proxy settings by its own or just use the system settings.
But thanks. I think this will be fixed in the future and until then, your idea is a nice workaround. I just don't want to install any file sync tool as thats not wanted here at the place but could sync files manually. :-)

Greets, Jürgen

**Comment by laurent22** [comment]
Implemented there: https://github.com/laurent22/joplin/commits/proxy_test but there are issues with timeout when the provided proxy URL is incorrect (the requests never time out) so that still needs to be fixed. The underlying network lib unfortunately doesn't allow controlling timeout properly so not sure what the fix is at this point.

**Comment by dimpflmoser** [comment]
Excellent, is there a binary / AppImage to test it?
The effect of requests that don't time out also exist today if you run the App behind a proxy, so it's no step backwards only one forward.

**Comment by laurent22** [comment]
No binary yet as I consider it's not stable enough for release. If you think it could be use, I can add it to the next release anyway and mark it as "Alpha" for instance.

**Comment by dimpflmoser** [comment]
It's definitely a valuable feature. Currently I use Android-Only (NextCloud) Notes by Stefan Niedermann, however it's markdown capabilities are rather limited. On the other hand editing Notes, especially migrating my existing notes on your Android client is a pain, so I would like to do it on the desktop. At work I'm caught behind a firewall and a proxy I cannot use it right now.

Sure a fallback if the proxy is not reachable would be great, however I'd regard that as the second step.

**Comment by dimpflmoser** [comment]
There have been a few releases since the last update on this issue, however I haven't seen an option in those updates. Is there an option to test to connect-via-proxy feature without compilation from source?

**Comment by laurent22** [comment]
The branch I had seemed to be working, at least with the local Squid proxy I had setup. But I've later tried with a regular proxy on the web and it didn't work. I'm not sure it works at all actually so it won't released for now.

**Comment by dimpflmoser** [comment]
What a shame, anyway thanks for your effort. - Still waiting (impatiently). :-)

**Comment by janLo** [comment]
This must be a regression as it seemed to have worked a while ago.

I use joplin behind a proxy. It synchronized a while ago (to nextcloud) but today I discovered that it stopped. This renders it completely useless for me. Other electron-apps (like Signal, Spotify, ..) work fine in this environment.

**Comment by bhlevca** [comment]
It appears there is a way to force node.js apps to go through a proxy with **global-tunnel-ng**

```
var globalTunnel = require('global-tunnel-ng');
 
globalTunnel.initialize({
    host: '10.0.0.10',
    port: 8080,
    proxyAuth: 'userId:password', // optional authentication
    sockets: 50 // optional pool size for each http and https
});
```

[Link](https://www.npmjs.com/package/global-tunnel-ng)

Perhaps this is the way to go.

**Comment by dimpflmoser** [comment]
Am I wrong to believe this issue has been abandoned?

**Comment by bhlevca** [comment]
It is not OneDrive only, all sync methods fail behind proxy

**Comment by bellma101** [comment]
I'm also eagerly watching this issue, hoping for a fix. I sync to Nextcloud, and I have to use a proxy to hit it.

**Comment by laurent22** [comment]
I could not find a satisfying solution so for now it's more or less on hold. Main issue is that if connection or anything doesn't work with the proxy, it's not possible to get proper errors (connection never times out). So often I don't know if it's the app or the proxy or something else that's not working.

Because in-depth work and specialist knowledge is needed for this feature, perhaps a bounty or some form of crowdfunding would help find someone to look into it and submit a pull request.

**Comment by bhlevca** [comment]
I posted in one of the forums a request and a  proposal that uses node.js
capab ility to work with a proxy. Joplin in not the only node.js app that
has this issue. Others have looked into this already. On the Signal  for
desktop is an entire thread  about it.

**Comment by dimpflmoser** [comment]
@laurent22 As far as I know it's the same issue with firefox: If your proxy accepts your request but doesn't deliver data Firefox hangs forever. Of course this is crap, but I guess it's the price we have to pay to have a crap proxy.

**Comment by think-nice-things** [comment]
yeah, I'm also "eagerly" watching this topic.
Didn't succed in building the proxy test version myself to test.
If someone would provide a binary test version (windows) I'd be happy to do some beta testing :-)

**Comment by iliuweiming** [comment]
Still...

**Comment by 0x7878** [comment]
So sad... I'm using Joplin on my mobile phone and would be really glad If i could using it behind a Corporate Proxy. So please fix it.

**Comment by theslash** [comment]
Any news on this? It's the one feature that's missing for me

**Comment by StefanTT** [comment]
Too bad that there is no solution for the proxy problem. Otherwise Joplin would be nearly perfect for my needs. Unfortunately a show stopper for me.

**Comment by nikosvaggalis** [comment]
That's true.I can't access my notes from work because I'm behind an NTLM proxy.Isn't there a work around like with using cntlm, or does Jopplin not work with a proxy at all,that is, not even accept a proxy as a an option?

**Comment by StefanTT** [comment]
I tried the typical environment variables for setting a proxy without success. And I think it simply does not work as the comments above in this thread state.

The use of an external sync tool is possible, of course.

**Comment by kenden** [comment]
To make this bug easier to find from google:
the error message received when trying to sync joplin with dropbox on a mac is:
```
Could not authorise application:

request to https://api.dropboxapi.com/oauth2/token failed, reason: connect ETIMEDOUT 170.135.66.9:443

Please try again.
```

![image](https://user-images.githubusercontent.com/480586/57449908-64317a80-725d-11e9-8c11-66351d35e093.png)

**Comment by g4c** [comment]
Hi, thanks for the app :-)

Also missing the proxy feature.

As a workaround I use sockscap64 on Windows. It's also portable.

**Comment by jmleroy** [comment]
I also have this problem at my office when working behind a corporate proxy.
I thought I could bypass the problem by using instead the filesystem sync on the joplin folder in my nextcloud folder, but my Joplin instances don't seem to be synced.

**Comment by 0x7878** [comment]
@jmleroy I made a similar experience with syncing on the Filesystem, mainly when you use another device that syncs normally.

@g4c Are you behind a Corporate socks proxy?  I mean, I know Sockscap, but it didn't work on HTTP Proxies  (at least not for me).

**Comment by kleinfelter** [comment]
OK.  If sync doesn't work through a proxy, how well does Joplin work with epically delayed sync?

e.g. If I use Joplin at work (with a proxy, and sync fails) one day, and I use Joplin at home the next day (and sync works),  what sorts of problems am I going to run into:

1. With no concurrent updates to the same file?
2. With concurrent updates to the same file?

**Comment by g4c** [comment]
@0x7878 Yes, corporate http proxy on port 80, no authorisation. Will give you more details how to configure once in the office.

**Comment by g4c** [comment]
@0x7878 Ok, that's easy. I did it long time ago and completely forgot - just using daily and that's it.
You just need to click Proxy to start Proxy Manager. Then just insert row and fill Proxy Address, Port (80), SOCKS Type (choose HTTP). Good luck!

**Comment by g4c** [comment]
Nope, it has not been fixed and I'm still waiting for a solution not requiring workarounds (using third party software)
Regards 
GC

**Comment by kenden** [comment]
Related/duplicate: https://github.com/laurent22/joplin/issues/164

**Comment by indirection** [comment]
Again for another duckduckgo.com (or google) search result, my error message:

FetchError: request to https://public.dm.files.1drv.com failed, reason: unable to get local issuer certificate
    at ClientRequest.<anonymous> (/Applications/Joplin.app/Contents/Resources/app/node_modules/node-fetch/index.js:133:11)

**Comment by indirection** [comment]
Thank you for Joplin! 

I've tried every work around that I could find. 

Completed: 2020-04-06 8:27 PM
Last error: FetchError: request to https://public.dm.files.1drv.com/<snip>/version.txt failed, reason: unable to get local issuer certificate

This example is behind the Cisco Umbrella proxy.

**Comment by bhlevca** [comment]
As far as I know the code to work with proxy has not been implemented yet.
I posted some suggedtions and someone has succeded to do it in a kludgy
way. The promlem is that node.js is not too friendly with proxies.

Bogdan

**Comment by bhlevca** [comment]
> Hey there, it looks like there has been no activity on this issue recently. Has the issue been fixed, or does it still require the community's attention? This issue may be closed if no further activity occurs. You may comment on the issue and I will leave it open. Thank you for your contributions.

As far as I know no work has been done on this. I think that it needs a fix not a closure.

**Comment by wangyongcong** [comment]
I have the same problem on MacOS. Joplin seems not respect system proxy setting when sync to Dropbox or checking update. This problem make Joplin hard to use in some country.

**Comment by zicklag** [comment]
If you use Nextcloud for the note sync, an easy solution that bypasses the proxy problem is to install the [Nextcloud Desktop client](https://nextcloud.com/install/#install-clients) and synchronize a Nextcloud folder to your local drive, then use the Joplin local filesystem synchronization mode.

**Comment by gibboni** [comment]
This needs a proper fix. 
Lack of proxy support prevented me using Joplin so I had to move on and use other editors for note taking. But I hope this is some day fixed so I could use it again.

**Comment by bhlevca** [comment]
Still hoping for a fix ...

**Comment by g4c** [comment]
Actually it's even worse. I installed the new version (portable) yesterday and the nasty workaround with sockscap64 stopped working. Joplin just hangs and that's it. I cannot move to the previous version as the db has been migrated to the new format 😭
When I disconnect vpn works ok. 
Guys, please upvote or advise how to raise priority of this issue...

**Comment by nokoa77** [comment]
@g4c argh I wanted to try also sockscap on Monday at work ! 
This feature is really missing, it make me very difficult to sync when I’m at work; I have to switch network several times a day to make it work and to always have an alternative WiFi network working in parallel of my enterprise WiFi network !! It’s really annoying.
And the « local sync + nextcloud folder » alternative doesn’t work : the sync is partial, it’s explained somewhere else in the forum ; local sync is not compatible with multi device sync.
I hope this feature can be implement even in alpha state if it can work on some situations

**Comment by nokoa77** [comment]
@gibboni as the OP, could you please rename the issue « sync fails when behind a proxy » ? As it is true for all OS ? Thanks

**Comment by g4c** [comment]
@nokoa77 Thanks for saving me stress with local sync I was planning to test. Now I know it's poinless 🙁

**Comment by g4c** [comment]
@gibboni yes, @nokoa77 is right, I use nextloud. Also add windows 10 please. Maybe this will make this thread more visible.

**Comment by zicklag** [comment]
> I use nextloud.

I am behind a proxy and I use Nextcloud's desktop client to synchronize the files to the local filesystem and then tell Joplin to use the Nextcloud-synced dir. It works really well, and Nextcloud sync supports the proxy. Definitely agree that this is in the workaround category and it only works with nextcloud which is unfortunate.

**Comment by nokoa77** [comment]
Hi @zicklag thank your testimonial, and sorry @g4c if I misinformed you by ignorance. In fact I tried for several weeks to sync using "local sync + nextcloud folder", and I counldn't make it work. I remember I saw in the joplin forum that it was a known bug, but apparently I was mistaken, maybe my conf was bad. I will try again. If @g4c you are more successfull than me, I'm interested !!

**Comment by g4c** [comment]
@nokoa77 no problem. I will let you know if any progress. But, honestly, I'm a bit tired by this waiting for any feedback with some kind of info and preferably deadline. I appreciate all the work invested in this app, really. But maybe it's time to search for some other solution. I've never faced any app having this kind of network problem. Keeping nextcloud folder synced locally opens new opportunities for using other apps.

**Comment by zicklag** [comment]
Thank you kind @stale bot, but we'd rather keep replying to this issue than let it get closed. :wink:

**Comment by nokoa77** [comment]
Hi, I confirm this feature is still an open issue for Joplin to work properly on all configurations, including when one is using it "at work" in a corporate company where proxy is the standard. There seems to be workarounds (like 1/ sync on file and 2/ sync this file with nextcloud or dropbox...) but it is not very efficient : I tried for several weeks and had to stop because of too many sync errors and fails.

**Comment by onatbas** [comment]
Hey guys, I appreciate the efforts put into Joplin, I truly do! But at the same time I can't believe how proxy support can be dismissed for so long. I believe you're underestimating the amount of people who direly needs proxy support.  This is a major hurdle for some (including me.)

**Comment by laurent22** [comment]
It's not a ignored, I've spent quite a bit of time tried to get it working but failed. I'd gladly accept a pull request for it but it would have to be rock solid or supported by a company or very motivated contributor, because I won't be able to fix any issue with it (setting up a proxy is too much hassle and when things don't work it's very hard to debug).

**Comment by g4c** [comment]
@laurent22 Thanks for the answer and all your work! I believed it should be more about getting new customers or companies. From my perspective this is basic functionally for every bigger company and there is also some competition on the market. Just don't understand why a company should insist on adding proxy? They just evaluate what's on the market and how it suits their needs. If they consider Joplin among others and find out it cannot work in their environment/infrastructure they just give up and choose something different instead. This is how it works based on my experience. But, of course, this is your business and your decision.

**Comment by gibboni** [comment]
@laurent22, have you tested global-agent? It works transparently to tunnel modules that use node's default http.request() method as well as the request module. It was made to replace the older global-tunnel library.

**Comment by vvvlad** [comment]
Any update about this topic, is there any plan to resolve it? Any workaround for MacOS?
Really hoped I could switch to Joplin but this is a deal breaker as I have proxy at work.

**Comment by jbockle** [comment]
@laurent22 as an alternative with file system synchronization and github, maybe joplin can expose post synchronization functionality where a user can specify a command/arguments to it so we can script a call to commit/push.  This could also open up additional use cases as well, such as publishing to a wiki/blog.

**Comment by kleinfelter** [comment]
I wonder if this could be accomplished via a transparent, local WebDAV proxy?  

Instead of configuring Joplin to sync via WebDAV to https://my-real-server/webdav it would be configured to sync with https://localhost/webdav. Then one would create a program to run on the local computer which presents a WebDAV interface, while silently relaying requests via the upstream corporate proxy.

**Comment by tiancheng2000** [comment]
As a workaround, I figured it out by configuring Tunnel mode on my Proxy Client (CFW in my case).
So maybe this issue can be ranked a lower priority?

**Comment by nokoa77** [comment]
> As a workaround, I figured it out by configuring Tunnel mode on my Proxy Client (CFW in my case).
> So maybe this issue can be ranked a lower priority?
Hi,
I'm not sure it is a solution for all situations, for instance in my situation (syncing with nextcloud on https), syncing on localhost won't work even via ssh tunneling :-(

**Comment by janLo** [comment]
There is at least signal, super-productivity and vscode that I know work with a proxy and use electron in some way. Maybe someone can have a look how these achieve this.

**Comment by linusguan** [comment]
Can anyone tell me the endpoint URL for synching with Joplin Cloud? Maybe I could ask our IT to set up bypass for it at the proxy side. Thanks!
@laurent22 maybe?

**Comment by laurent22** [comment]
That would api.joplincloud.com

**Comment by vbo75** [comment]
Hi,
Great job for new Joplin 2.6 version but PLEASE, add proxy support.
It's really a "Must Have" feature for business context where proxy is the standard. 
Thanks a lot !

**Comment by laurent22** [comment]
If I'm not mistaken the majority of users who need this feature need it because they are behind a company firewall.

In that case, consider asking your company to either fund the project, or post a bounty that will motivate developers. Or to allocate one of their own developers on the feature.

There's no need to keep posting to say it's a super important feature if no-one is willing to work on it.

**Comment by bhlevca** [comment]
I don't think and company which has a firewall would fund anything like that. They would use software that is "allowed" in the company, usually Google or Microsoft. On the contrary, they would see that as a breach in security.

If cost is the issue and it is  not technical, perhaps the solution would be crowdfunding, some of the links below could be a good start:
https://www.gofundme.com/en-ca/c/crowdfunding
https://www.kickstarter.com/

I am convinced that some if not all all on this email list would contribute with something, you just need to estimate the costs and come up with a project plan.

**Comment by jasonwilliams** [comment]
Hey @laurent22 I'm happy to look into this if you can point me in the right direction. I understand it's NodeJS trying to make a request but it could hang right? Im sure its possible to add a timeout in there at this point.  My guess is somewhere around [here](https://github.com/laurent22/joplin/blob/9684b38f7e8c9a304a541039db4f4e2d03ba9f77/packages/lib/shim-init-node.js#L368-L480)

From doing some digging around https://github.com/delvedor/hpagent seems like the best choice, 1 because it has [timeout support](https://github.com/delvedor/hpagent/pull/66) and it also looks to be maintained ([http-node-proxy library](https://github.com/TooTallNate/node-https-proxy-agent) isn't anymore).

I'm assuming it should be possible to add a configuration section for people to enter proxy information, or fallback to checking environment variables. These can be passed into the http(s) agent, there should also be a chance for it to fallback to requesting without a proxy should that fail (this is useful for those who go in and out of corporate networks).

Is there any way this can be tested by one of us? My company would not contribute financially but I can be given time for open source contributions and can test it with our firewall.

There's also many on this thread willing to test should we get some dev build out.

**Comment by jasonwilliams** [comment]
I'm pleased to say I have a working PR, details are here: https://github.com/laurent22/joplin/pull/6537
It would be helpful if some people here could check it out and try it in their environment then report back on the PR.

**Comment by IcePlains** [comment]
Thank you @jasonwilliams

Any chance of a socks proxy support as well?

**Comment by jasonwilliams** [comment]
@agiz10 the dependency I used only supports http(s) so it would require some work adding in something like https://github.com/JoshGlazebrook/socks
You could follow my pr to see what I did, should be easy

**Comment by echo672** [comment]
> If you're using Windows, you can use [proxifier](https://www.proxifier.com/) to redirect Joplin's connection via proxy. For Linux, use proxychains, run it like this: `proxychains ~/.joplin/Joplin.AppImage`. Still, I'd prefer a native proxy support that doesn't require 3rd party software :)

I use this method solving this problem in debian 

export http_proxy="http://ip:port"
export https_proxy="https:ip:port"
.joplin/Joplin.AppImage

**Comment by bernd-wechner** [comment]
The equivalent in Powershell would be:

```
$env:http_proxy = 'http://127.0.0.1:port'
$env:https_proxy = 'https://127.0.0.1:port'
.\JoplinePortable.exe
```

Alas it doesn't work with a SOCKS5 proxy :-(.
